File: packages/python/tileguard/utils/pbf_decoder.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

GEOM_TYPES = {0: "Unknown", 1: "Point", 2: "LineString", 3: "Polygon"}


@dataclass
class PbfReader:
    data: bytes
    pos: int = 0

    @property
    def eof(self) -> bool:
        return self.pos >= len(self.data)

    def read_varint(self) -> int:
        result = 0
        shift = 0
        while self.pos < len(self.data):
            byte = self.data[self.pos]
            self.pos += 1
            result |= (byte & 0x7F) << shift
            if not byte & 0x80:
                return result
            shift += 7
            if shift > 63:
                raise ValueError("Invalid varint: too many bytes")
        raise ValueError("Unexpected end of protobuf while reading varint")

    def read_svarint(self) -> int:
        value = self.read_varint()
        return (value >> 1) ^ -(value & 1)

    def read_bytes(self) -> bytes:
        size = self.read_varint()
        self.ensure(size)
        value = self.data[self.pos:self.pos + size]
        self.pos += size
        return value

    def read_string(self) -> str:
        return self.read_bytes().decode("utf-8")

    def read_float(self) -> float:
        import struct
        self.ensure(4)
        value = struct.unpack_from("<f", self.data, self.pos)[0]
        self.pos += 4
        return value

    def read_double(self) -> float:
        import struct
        self.ensure(8)
        value = struct.unpack_from("<d", self.data, self.pos)[0]
        self.pos += 8
        return value

    def skip(self, wire_type: int) -> None:
        if wire_type == 0:
            self.read_varint()
        elif wire_type == 1:
            self.ensure(8)
            self.pos += 8
        elif wire_type == 2:
            size = self.read_varint()
            self.ensure(size)
            self.pos += size
        elif wire_type == 5:
            self.ensure(4)
            self.pos += 4
        else:
            raise ValueError(f"Unsupported protobuf wire type {wire_type}")

    def ensure(self, size: int) -> None:
        if self.pos + size > len(self.data):
            raise ValueError("Unexpected end of protobuf")


def decode_mvt(data: bytes) -> dict[str, Any]:
    reader = PbfReader(data)
    layers: dict[str, Any] = {}
    while not reader.eof:
        tag = reader.read_varint()
        field = tag >> 3
        wire = tag & 7
        if field == 3 and wire == 2:
            layer = _decode_layer(reader.read_bytes())
            if layer["name"]:
                layers[layer["name"]] = layer
        else:
            reader.skip(wire)
    return {"layers": layers}


def _decode_layer(data: bytes) -> dict[str, Any]:
    reader = PbfReader(data)
    layer = {"name": "", "version": None, "extent": 4096, "keys": [], "values": [], "features": []}
    while not reader.eof:
        tag = reader.read_varint()
        field = tag >> 3
        wire = tag & 7
        if field == 15 and wire == 0:
            layer["version"] = reader.read_varint()
        elif field == 1 and wire == 2:
            layer["name"] = reader.read_string()
        elif field == 2 and wire == 2:
            layer["features"].append(_decode_feature(reader.read_bytes()))
        elif field == 3 and wire == 2:
            layer["keys"].append(reader.read_string())
        elif field == 4 and wire == 2:
            layer["values"].append(_decode_value(reader.read_bytes()))
        elif field == 5 and wire == 0:
            layer["extent"] = reader.read_varint()
        else:
            reader.skip(wire)

    for feature in layer["features"]:
        feature["properties"] = _hydrate_properties(feature.pop("tags"), layer["keys"], layer["values"])
        feature["geometry"] = _decode_geometry(feature.pop("geometry_commands"), feature["type"])
    return layer


def _decode_feature(data: bytes) -> dict[str, Any]:
    reader = PbfReader(data)
    feature = {"id": None, "type": 0, "geometry_type": "Unknown", "tags": [], "geometry_commands": []}
    while not reader.eof:
        tag = reader.read_varint()
        field = tag >> 3
        wire = tag & 7
        if field == 1 and wire == 0:
            feature["id"] = reader.read_varint()
        elif field == 2 and wire == 2:
            feature["tags"] = _read_packed_varints(reader.read_bytes())
        elif field == 3 and wire == 0:
            feature["type"] = reader.read_varint()
            feature["geometry_type"] = GEOM_TYPES.get(feature["type"], "Unknown")
        elif field == 4 and wire == 2:
            feature["geometry_commands"] = _read_packed_varints(reader.read_bytes())
        else:
            reader.skip(wire)
    return feature


def _decode_value(data: bytes) -> Any:
    reader = PbfReader(data)
    value = None
    while not reader.eof:
        tag = reader.read_varint()
        field = tag >> 3
        wire = tag & 7
        if field == 1 and wire == 2:
            value = reader.read_string()
        elif field == 2 and wire == 5:
            value = reader.read_float()
        elif field == 3 and wire == 1:
            value = reader.read_double()
        elif field == 4 and wire == 0:
            value = reader.read_svarint()
        elif field == 5 and wire == 0:
            value = reader.read_varint()
        elif field == 6 and wire == 0:
            value = reader.read_svarint()
        elif field == 7 and wire == 0:
            value = bool(reader.read_varint())
        else:
            reader.skip(wire)
    return value


def _hydrate_properties(tags: list[int], keys: list[str], values: list[Any]) -> dict[str, Any]:
    props = {}
    for index in range(0, len(tags), 2):
        key_index = tags[index]
        value_index = tags[index + 1]
        if key_index < len(keys) and value_index < len(values):
            props[keys[key_index]] = values[value_index]
    return props


def _read_packed_varints(data: bytes) -> list[int]:
    reader = PbfReader(data)
    values = []
    while not reader.eof:
        values.append(reader.read_varint())
    return values


def _decode_geometry(commands: list[int], geom_type: int) -> list[Any]:
    x = 0
    y = 0
    index = 0
    current = []
    parts = []
    while index < len(commands):
        command_integer = commands[index]
        index += 1
        command = command_integer & 0x7
        count = command_integer >> 3
        if command in (1, 2):
            if command == 1 and current:
                parts.append(current)
                current = []
            for _ in range(count):
                x += _zig_zag_decode(commands[index])
                y += _zig_zag_decode(commands[index + 1])
                index += 2
                current.append({"x": x, "y": y})
        elif command == 7:
            for _ in range(count):
                if current:
                    current.append(dict(current[0]))
        else:
            raise ValueError(f"Unsupported geometry command {command}")
    if current:
        parts.append(current)
    if geom_type == 1:
        return [point for part in parts for point in part]
    return parts


def _zig_zag_decode(value: int) -> int:
    return (value >> 1) ^ -(value & 1)

File: legacy/js/src/utils/pbf-decoder.js
const GEOM_TYPES = {
  0: 'Unknown',
  1: 'Point',
  2: 'LineString',
  3: 'Polygon'
};

export class PbfReader {
  constructor(buffer) {
    this.buf = Buffer.isBuffer(buffer) ? buffer : Buffer.from(buffer);
    this.pos = 0;
    this.length = this.buf.length;
  }

  eof() {
    return this.pos >= this.length;
  }

  readVarint() {
    let result = 0;
    let shift = 0;
    while (this.pos < this.length) {
      const byte = this.buf[this.pos++];
      if (shift < 28) {
        result += (byte & 0x7f) << shift;
      } else {
        result += (byte & 0x7f) * 2 ** shift;
      }
      if ((byte & 0x80) === 0) return result;
      shift += 7;
      if (shift > 63) throw new Error('Invalid varint: too many bytes');
    }
    throw new Error('Unexpected end of protobuf while reading varint');
  }

  readSVarint() {
    const n = this.readVarint();
    return (n >>> 1) ^ -(n & 1);
  }

  readDouble() {
    this.ensure(8);
    const value = this.buf.readDoubleLE(this.pos);
    this.pos += 8;
    return value;
  }

  readFloat() {
    this.ensure(4);
    const value = this.buf.readFloatLE(this.pos);
    this.pos += 4;
    return value;
  }

  readBytes() {
    const len = this.readVarint();
    this.ensure(len);
    const value = this.buf.subarray(this.pos, this.pos + len);
    this.pos += len;
    return value;
  }

  readString() {
    return this.readBytes().toString('utf8');
  }

  skip(wireType) {
    if (wireType === 0) {
      this.readVarint();
      return;
    }
    if (wireType === 1) {
      this.ensure(8);
      this.pos += 8;
      return;
    }
    if (wireType === 2) {
      const len = this.readVarint();
      this.ensure(len);
      this.pos += len;
      return;
    }
    if (wireType === 5) {
      this.ensure(4);
      this.pos += 4;
      return;
    }
    throw new Error(`Unsupported protobuf wire type ${wireType}`);
  }

  ensure(len) {
    if (this.pos + len > this.length) {
      throw new Error('Unexpected end of protobuf');
    }
  }
}

export function decodeMvt(buffer) {
  const reader = new PbfReader(buffer);
  const layers = {};

  while (!reader.eof()) {
    const tag = reader.readVarint();
    const field = tag >> 3;
    const wire = tag & 7;

    if (field === 3 && wire === 2) {
      const layer = decodeLayer(reader.readBytes());
      if (layer.name) layers[layer.name] = layer;
    } else {
      reader.skip(wire);
    }
  }

  return { layers };
}

function decodeLayer(bytes) {
  const reader = new PbfReader(bytes);
  const layer = {
    name: '',
    version: null,
    extent: 4096,
    keys: [],
    values: [],
    features: []
  };

  while (!reader.eof()) {
    const tag = reader.readVarint();
    const field = tag >> 3;
    const wire = tag & 7;

    if (field === 15 && wire === 0) layer.version = reader.readVarint();
    else if (field === 1 && wire === 2) layer.name = reader.readString();
    else if (field === 2 && wire === 2) layer.features.push(decodeFeature(reader.readBytes(), layer));
    else if (field === 3 && wire === 2) layer.keys.push(reader.readString());
    else if (field === 4 && wire === 2) layer.values.push(decodeValue(reader.readBytes()));
    else if (field === 5 && wire === 0) layer.extent = reader.readVarint();
    else reader.skip(wire);
  }

  for (const feature of layer.features) {
    feature.properties = hydrateProperties(feature.tags, layer.keys, layer.values);
    feature.geometry = decodeGeometry(feature.geometryCommands, feature.type);
    delete feature.tags;
    delete feature.geometryCommands;
  }

  return layer;
}

function decodeFeature(bytes) {
  const reader = new PbfReader(bytes);
  const feature = {
    id: undefined,
    type: 0,
    geometryType: 'Unknown',
    tags: [],
    geometryCommands: []
  };

  while (!reader.eof()) {
    const tag = reader.readVarint();
    const field = tag >> 3;
    const wire = tag & 7;

    if (field === 1 && wire === 0) feature.id = reader.readVarint();
    else if (field === 2 && wire === 2) feature.tags = readPackedVarints(reader.readBytes());
    else if (field === 3 && wire === 0) {
      feature.type = reader.readVarint();
      feature.geometryType = GEOM_TYPES[feature.type] || 'Unknown';
    } else if (field === 4 && wire === 2) feature.geometryCommands = readPackedVarints(reader.readBytes());
    else reader.skip(wire);
  }

  return feature;
}

function decodeValue(bytes) {
  const reader = new PbfReader(bytes);
  let value = null;

  while (!reader.eof()) {
    const tag = reader.readVarint();
    const field = tag >> 3;
    const wire = tag & 7;

    if (field === 1 && wire === 2) value = reader.readString();
    else if (field === 2 && wire === 5) value = reader.readFloat();
    else if (field === 3 && wire === 1) value = reader.readDouble();
    else if (field === 4 && wire === 0) value = reader.readSVarint();
    else if (field === 5 && wire === 0) value = reader.readVarint();
    else if (field === 6 && wire === 0) value = reader.readSVarint();
    else if (field === 7 && wire === 0) value = Boolean(reader.readVarint());
    else reader.skip(wire);
  }

  return value;
}

function hydrateProperties(tags, keys, values) {
  const props = {};
  for (let i = 0; i < tags.length; i += 2) {
    const key = keys[tags[i]];
    if (key !== undefined) props[key] = values[tags[i + 1]];
  }
  return props;
}

function readPackedVarints(bytes) {
  const reader = new PbfReader(bytes);
  const values = [];
  while (!reader.eof()) values.push(reader.readVarint());
  return values;
}

function decodeGeometry(commands, type) {
  let x = 0;
  let y = 0;
  let i = 0;
  let current = [];
  const parts = [];

  while (i < commands.length) {
    const commandInteger = commands[i++];
    const command = commandInteger & 0x7;
    const count = commandInteger >> 3;

    if (command === 1 || command === 2) {
      if (command === 1 && current.length) {
        parts.push(current);
        current = [];
      }
      for (let c = 0; c < count; c++) {
        x += zigZagDecode(commands[i++]);
        y += zigZagDecode(commands[i++]);
        current.push({ x, y });
      }
    } else if (command === 7) {
      for (let c = 0; c < count; c++) {
        if (current.length) current.push({ ...current[0] });
      }
    } else {
      throw new Error(`Unsupported geometry command ${command}`);
    }
  }

  if (current.length) parts.push(current);
  if (type === 1) return parts.flat();
  return parts;
}

function zigZagDecode(n) {
  return (n >>> 1) ^ -(n & 1);
}
